[WiP] Set propagation to rslave for hostPath volume mounts

ivan4th opened the pull request:
Based on #31504, but supports only rslave, adds support for CRI /
dockershim and displays warnings of Docker doesn't support mount
propagation options.

Not to be merged till kubernetes/community#193 discussion is resolved.
TODO: e2e test.

Commit:
Set propagation to rslave for hostPath mounts

Based on #31504, but supports only rslave, adds support for CRI /
dockershim and displays warnings of Docker doesn't support mount
propagation options.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41683)
<!-- Reviewable:end -->

gurvindersingh commented on pkg/kubelet/dockershim/helpers.go:
```diff
@@ -108,29 +111,35 @@ func extractLabels(input map[string]string) (map[string]string, map[string]strin
+// is a comma-separated list of the following strings:
+// 'ro', if the path is read only
+// 'Z', if the volume requires SELinux relabeling
+// propagation mode such as 'rshared'
```
May be you can mention `rslave` here, as that's what this patch adds.

gurvindersingh commented on pkg/kubelet/dockertools/docker_manager.go:
```diff
@@ -545,23 +545,30 @@ func makeEnvList(envs []kubecontainer.EnvVar) (result []string) {
+// is a comma-separated list of the following strings:
+// 'ro', if the path is read only
+// 'Z', if the volume requires SELinux relabeling
+// propagation mode such as 'rshared'
```
same as above

ivan4th commented:
Current e2e failures are caused by this:
```E0218 03:15:56.779026    3290 remote_runtime.go:203] StartContainer "cb3622a4e409f66d78e5e06a9d5b296fd41b3ccebfd7e2612f4399e00b517b5c" from runtime service failed: rpc error: code = 2 desc = failed to start container "cb3622a4e409f66d78e5e06a9d5b296fd41b3ccebfd7e2612f4399e00b517b5c": Error response from daemon: linux mounts: Path /var/log is mounted on / but it is not a shared or slave mount.
E0218 03:15:56.780689    3290 kuberuntime_manager.go:715] container start failed: rpc error: code = 2 desc = failed to start container "cb3622a4e409f66d78e5e06a9d5b296fd41b3ccebfd7e2612f4399e00b517b5c": Error response from daemon: linux mounts: Path /var/log is mounted on / but it is not a shared or slave mount.: Start Container Failed
E0218 03:15:56.781042    3290 pod_workers.go:182] Error syncing pod f76dc21cc44d2ff38117b64cc043dd01 ("kube-proxy-gke-e2e-gke-agent-pr-83--default-pool-3a973fae-3z86_kube-system(f76dc21cc44d2ff38117b64cc043dd01)"), skipping: failed to "StartContainer" for "kube-proxy" with rpc error: code = 2 desc = failed to start container "cb3622a4e409f66d78e5e06a9d5b296fd41b3ccebfd7e2612f4399e00b517b5c": Error response from daemon: linux mounts: Path /var/log is mounted on / but it is not a shared or slave mount.: "Start Container Failed"```

Apparently newer dockers (like 1.13.x) check whether the source path is a mount and bindmount it on itself if it isn't. Looks like older dockers don't do this. This can be worked around by doing the checks on kubelet/dockershim side, but the problem is (already discussed to some extent but I don't recall any definite solution being found) that docker daemon / containerd may reside in its own mount namespace and thus checking the mounts on the hosts may still mean that they're not there from docker's PoV. A possible smoke test for this is checking current mounts and mount propagation to the Docker using e.g. busybox container at kubelet/dockershim startup but this is ... hacky. Any thoughts on a good way to deal with this? Sorry if it was already mentioned in some of discussions but I missed it.

ivan4th commented:
@gurvindersingh thanks for pointing out the problems in comments, fixed them

gurvindersingh commented:
@euank @majewsky @lucab @dchen1107 wondering if you guys can go over the current work from @ivan4th  in this PR and review it, so we can fix it with the comments and hopefully make it ready for merge before 1.6 code freeze.

Thanks!!

euank commented:
@ivan4th if the right way is to do a dryrun/smoketest contains, the pause container would be a better option than busybox.

There could also be kubelet code to find dockerd's pid, break into its mount namespace, and do the test from its perspective.

jsafrane commented:
> the test failure in e2e is because / is not a slave mount, but if the kubelet made it one (at least for docker's mountns) it would probably work... and maybe that's the right thing to do for the user.

That sounds dangerous. Somebody (system admin, distribution author) has decided that Docker should run in its own mount namespace with specific propagation mode and it's not up to Kubernetes to try to fix that.

IMO we should print some error and refuse to do slave or shared mounts instead of trying to hack the system. It would be better to check and report this misconfiguration during node startup and not when the first pod with HostPath is created, but it's IMO not a hard requirement.

jsafrane commented:
I checked major distros, always with the distribution docker (Fedora, CentOS, CoreOS) or the latest from docker.io (Debian, Ubuntu):

* CoreOS Container Linux 1353.1.0, Fedora 25, Debian stable (jessie), Ubuntu 16.04 LTS run with root mount as shared, which is propagated into docker (i.e. docker does not have its own mount namespace nor remounts the root). This PR should work there without any special config.
* RHEL7/CentOS7 runs with root as shared, however docker runs in its own mount namespace and sees root as slave. This could be fixed in a future update (no promises though).
* Debian wheezy (old stable) runs with root as private. That results in error in GCE tests we see here.

So, do we need to support wheezy? Can we change default private root to shared in GCE test infrastructure? Can we upgrade to jessie?

I'll check docker 1.13 and cryptic bind-mounts, but I did not notice anything bad in ubuntu 16.04 with docker community edition that reports itself as version 17.03.0-ce...

jsafrane commented:
Checked with Container Linux 1353.1 which ships with docker 1.13.1, I did not notice any bad bind mounts that would prevent `docker run  -v xxx:yyy:rslave` from working. @ivan4th, on what distro do you see these bind mounts?

jsafrane commented:
@euank @ivan4th, I've been thinking, what kind of autodetection could we do on kubelet startup? There may be many (bin)mounts on the system, either on the host or in docker mount namespace, which one should we check if it's shared? We know only at a HostPath pod creation which path mount(s) need to be shared/slave and which can stay private. Does check of `/` in docker mount namespace tell us anything?

vishh commented:
It is not OK for any arbitrary pod to be able to introduce mounts into the host's mount namespace. I see the need for this feature. At the least it has to be gated by a Pod Security Policy field and only exposed to certain trusted pods.

cc @saad-ali @euank

euank commented:
@vishh this is about rslave, not rshared.

This change does not allow a pod to change the host's mount namespace anymore than it already can.

vishh commented:
@euank sorry, I was focusing mainly on the overall use case. Given what you said, I'm curious what the use case is for this PR?

euank commented:
@vishh A couple use-cases were already mentioned in the original issue, such as https://github.com/kubernetes/kubernetes/pull/31504#issuecomment-273539727

gurvindersingh commented:
@ivan4th any update on this one.

ivan4th commented:
I've updated the PR for current master. Still, I don't know how to solve the current Docker problem. Indeed it sounds like tweaking docker daemon's mount ns without user's consent is scary. Failing loudly is not quite an option either because we're changing the default behavior here, so we may break k8s for some users. Also, the CI env has still this / mount problem.

kfox1111 commented:
Is anyone working on shared mounts anymore? how do we ensure shared mounts are in 1.7.

gurvindersingh commented:
@ivan4th may be we can gate this setting using experimental flags and mentioned the docker mount settings required to make this option work. For CI issue, @jsafrane created #44389 which can possibly make the tests passing for this PR.

ivan4th commented:
@gurvindersingh yes, this sounds like a plan. I'll add the flag to kubelet in this PR, and let's wait for CI fix to land. Just in case -- here's why some installations specify `MountFlags` for docker.service: http://blog.oddbit.com/2015/01/18/docker-vs-privatetmp/

Commit:
Add kubelet option to enable rslave hostPath

The option is --experimental-enable-host-path-mount-propagation,
defaults to false.

ivan4th commented:
Added `--experimental-enable-host-path-mount-propagation` kubelet option, defaults to `false`.

jsafrane commented:
yet another implementation attempt: https://github.com/kubernetes/kubernetes/pull/46444
